### PR TITLE
(WIP) lazy log messages: Replacing or continuing #105

### PR DESCRIPTION
**Pull request by acodili-jg**
#### Removals

 - `boolean ALUtils.checkErrors(Supplier<String>)`

#### Additions

 - `Optional<String> ALUtils.consumeErrorMessage()`
 - `boolean ALUtils.logAndConsumeApplyError(String, int, String, int)`
 - `boolean ALUtils.logAndConsumeError(Object)`
 - `boolean ALUtils.logAndConsumeMultiApplyError(String[], int[], String, int)`
 - `boolean ALUtils.logAndConsumeSetError(String, String, String, float)`
 - `boolean ALUtils.logAndConsumeSetError(String, String, int, float)`
 - `boolean ALUtils.logAndConsumeSetError(String, int, String, float)`

#### Changes

 - `boolean ALUtils.errorApply(String, int, String, int)`
    - delegates to `boolean ALUtils.logAndConsumeApplyError(String, int, String, int)`
 - `boolean ALUtils.errorApply(String[], int[], String, int)`
    - delegates to `boolean ALUtils.logAndConsumeMultiApplyError(String[], int[], String, int)`
 - `boolean ALUtils.errorProperty(String, String, int, float)`
    - delegates to `boolean ALUtils.logAndConsumeSetError(String, String, int, float)`
 - `boolean ALUtils.errorProperty(String, int, String, float)`
    - delegates to `boolean ALUtils.logAndConsumeSetError(String, int, String, float)`
 - `boolean Cache.generate(Consumer<String>)`
    - changed signature to `boolean Cache.generate(LogBuilder)`
       - all member calls to the first parameter are replaced with an equivalent
    - the recent `boolean Cache.generate()` might supersede this method

<details>
<summary>Old content</summary>

 - ~~`checkErrors(Supplier<String>)` is renamed to `logErrors(Supplier<String>)`~~
    - ~~The rename better describes the method, though it seems that `ALUtils.checkErrors(String)` is based on Minecraft's `ALUtil.checkErrors(String)`~~
 - Replaced `checkErrors(Supplier<String>)`
 - Revert changes to `checkErrors(String)` and wrapped the string concat in a supplier
 - Rename Cache#generate(Consumer<String>) to Cache#generate(LogBuilder) ([2b7bf56](2b7bf5608874a56b40cc2408db64ad1b8f1c0965))
 - Replaced `Consumer<String>` loggers to `LogBuilder`, usually acquired using `Logger.atXxx()` or `Logger.atLevel(Level)`, loggers
    - Though `LogBuilder` is more similar to `Logger`, it could be more restrictive than `Consumer<String>` (e.g. `message -> { doSomething(message); LOGGER.log(message); }`).
 - Updated commented code
 - Some templated logging message lines' parameters are wrapped into a supplier
 - Replace streamed for-each logging into a multiline message sent in bulk
    - Benefit is that the stream and the operations are only performed when logging can happen (e.g. matching log level)
    - Behavioral changes, for a theoretical scenario where the logging level changes mid-iterating the stream:
       - Previous behavior would partially print the entries of the stream
       - Current implementation would either print the whole stream or none of it
</details>
 - and some more...

### CODE REVIEW

File: src/main/java/dev/thedocruby/resounding/Cache.java [hunk 1]
```diff
@@ -282,17 +285,17 @@ public static Branch growOctree(WorldChunk chunk, Branch root) {
...
-                    );
-                }
-        );
+        logger.log("{}", () -> {
+            return Registry.REGISTRIES.streamTags()
+                    .map(TagKey::registry)
+                    .map(RegistryKey::getValue)
+                    .map(Identifier::getPath)
+                    .collect(Collectors.joining("\n"));
+        });
```
Comment (mikenrafter) [inline src/main/java/dev/thedocruby/resounding/Cache.java]: While this code isn't of much consequence, given it's being refactored... Multiple maps have worse performance than a single iteration, and this degrades readability in this circumstance.

**Review by mikenrafter — CHANGES_REQUESTED** [review]
I haven't fully tested these changes, but most all seems well!

**Comment by mikenrafter** [comment]
I'll merge this once I get the cache system built.

File: src/main/java/dev/thedocruby/resounding/Cache.java (hunk 1 above)
Comment (acodili-jg) [reply src/main/java/dev/thedocruby/resounding/Cache.java]: I see. I will replace this with a single map or maybe find an alternative to streamTags().

**Comment by mikenrafter** [comment]
How are you addressing: [lambdas on the heap](https://stackoverflow.com/questions/27524445/does-a-lambda-expression-create-an-object-on-the-heap-every-time-its-executed) and [lambda closure](https://www.infoq.com/articles/Java-8-Lambdas-A-Peek-Under-the-Hood/)?

Do you think performance degradation when logging is worth it? If so, could we get some profiling done to know what kind of hit these closure-accessing lambdas cause?

\**not being condescending, genuinely curious to hear your take*

**Comment by acodili-jg** [comment]
These didn't come to my mind before being mentioned. Though most lambda-*fications* are mostly done to `ALUtils` (and code calling to its members), I was ~~over~~ thinking about limiting the number of lambda closures. Right now, I'm currently thinking about making an alternative to (not outright replace) `ALUtils.checkErrors(Consumer<String>)` as I'm aiming to use parameterized message logging to replace most lambdas that are connected to `ALUtils`.

<details>
<summary></summary>

> I had no access to the internet for two days. I did get to see this PR for a bit on that last day. This message totally didn't take a whole morning to reword.
</details>

**Comment by StartsMercury** [comment]
Inactive and outdated. I haven't checked in a while but I think it is better to create a new PR if this is still needed (and maybe start over since these aren't big changes) and close this one for now.

**Comment by acodili-jg** [comment]
Inactive and outdated. I haven't checked in a while but I think it is better to create a new PR if this is still needed (and maybe start over since these aren't big changes) and close this one for now.

**Pull request closed without merging**